### PR TITLE
move amp checkpoint state management to precision plugin

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7652 (splitting down diff)

Makes checkpoint saving/loading logic agnostic to precision being used.
This is just a refactor, no changes to the current behavior are intended.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]
LGTM. Just a question: could we maybe group those args under a precision subdict so that we don't pollute the global checkpoint namespace?

**Comment by awaelchli** [comment]
would be great but then we would have to deal with backward compatibility too 🙈 
can do a follow up :)

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 1]
```diff
@@ -127,3 +127,22 @@ def clip_grad_by_norm(self, optimizer: Optimizer, clip_val: Union[int, float]) -
...
+
+    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """
+        Called when Trainer is saving a checkpoint.
+
+        Args:
+            checkpoint: The full checkpoint dictionary before it gets dumped to a file.
+                Implementations of this hook can insert additional data into this dictionary.
+        """
+        pass
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/precision_plugin.py:148]: why not use the CheckpointHooks interface? https://github.com/PyTorchLightning/pytorch-lightning/blob/ca89a7f3443710a2313e97b91ac908867eb33be5/pytorch_lightning/core/hooks.py#L782

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/precision_plugin.py:148]: sure, I could add this in as a parent class if that's the intended use case

**Review by carmocca — APPROVED** [review]
Want to update the CHANGELOG for this?

**Review by ananthsub — APPROVED** [review]

**Pull request merged**